server/storage.ts

```diff
@@ -59,11 +61,68 @@ class DiskStorageEngine implements IStorageEngine {
 	}
 }
 
+interface IS3Options extends ICommonOptions {
+	bucket: string;
+	region: string;
+	accessKey: string;
+	secretKey: string;
+}
+
+class S3StorageEngine implements IStorageEngine {
+	public readonly uploadRoot: string;
+	private readonly options: IS3Options;
+
+	constructor(options: IS3Options) {
+		// Values copied via spread operator instead of being passed by reference
+		this.options = {
+			...options
+		};
+		this.uploadRoot = this.options.uploadDirectory;
```

Comment:
@petschekr No. Buckets are pretty static and the account in question should not even have permission to modify bucket-level settings in S3.